casper/src/test/scala/coop/rchain/casper/MultiParentCasperMergeSpec.scala

```diff
@@ -53,34 +61,161 @@ class MultiParentCasperMergeSpec extends FlatSpec with Matchers with Inspectors
   }
 
   it should "handle multi-parent blocks correctly when they operate on stdout" ignore effectTest {
-    def echoContract(no: Int) = s"""new stdout(`rho:io:stdout`) in { stdout!("Contract $no") }"""
-    val time                  = System.currentTimeMillis()
-    HashSetCasperTestNode.networkEff(genesis, networkSize = 2).use { nodes =>
-      val deploys = Vector(
-        ConstructDeploy.sourceDeploy(echoContract(1), timestamp = time + 1),
-        ConstructDeploy.sourceDeploy(echoContract(2), timestamp = time + 2)
-      )
-      for {
-        block0 <- nodes(0).addBlock(deploys(0))
-        block1 <- nodes(1).addBlock(deploys(1))
-        _      <- nodes(0).receive()
-        _      <- nodes(1).receive()
-        _      <- nodes(0).receive()
-        _      <- nodes(1).receive()
+    def echoContract(no: Int) =
+      Rho(s"""new stdout(`rho:io:stdout`) in { stdout!("Contract $no") }""")
+    merges(echoContract(1), echoContract(2), Rho("Nil"))
+  }
 
-        //multiparent block joining block0 and block1 since they do not conflict
-        multiparentBlock <- nodes(0).addBlock(deploys(1))
-        _                <- nodes(1).receive()
+  // TODO: Peek rows/column
+  // Note this skips pairs that lead to infinite loops
+  it should "handle multi-parent blocks correctly when they operate on volatile produce/consume pairs" in effectTest {
+    // Sends (linear sends)
+    val S0 = Rho("@0!(0)")
+    val S1 = Rho("@0!(1)")
+    // Repeats (persistent sends)
+    val R0 = Rho("@0!!(0)")
+    val R1 = Rho("@0!!(1)")
+    // For-s (linear receives)
+    val F_ = Rho("for (_ <- @0) { 0 }")
+    val F0 = Rho("for (@0 <- @0) { 0 }")
+    val F1 = Rho("for (@1 <- @0) { 0 }")
+    // Contracts (persistent receives)
+    val C_ = Rho("contract @0(id) = { 0 }")
+    val C0 = Rho("contract @0(@0) = { 0 }")
+    val C1 = Rho("contract @0(@1) = { 0 }")
+    //FIXME add missing cases for in-deploy COMM-s wherever there's a pair without X (a COMM)
+    //FIXME all `conflictsForNow` should eventually be replaced with `merges`
+    Map(
+      "!X !X"        -> conflictsForNow(S0, S0, Nil),
+      "!X !4"        -> conflictsForNow(S0, S0, F_),
+      "!X (!4)"      -> merges(S0, S1 | F_, Nil),
+      "!X !C"        -> conflictsForNow(S0, S1, C1),
+      "!X (!C)"      -> conflicts(S0, S1 | C_, Nil),
+      "!X 4X"        -> conflicts(S0, F_, Nil),
```

Comment:
TDDM: `If we can prove they don't match, then they can merge.`

are there any plans to address this case? I guess this would mean running the SpatialMatcher.

Reply:
I don't think running the SM will be necessary. No match === no COMM event and this is all the info we need and have

Reply:
This one conflicts because `F_` and `S0` do match

Reply:
that's the point: IF we can show that `!X` would not match with `F_` it's still fine to merge.

Reply:
Right, so we need to add `"!X 4X - no match"        -> conflictsForNow(S0, F1, Nil),`

Reply:
Notice also that using spatial matcher requires having actual data and continuations, not just hashes. So it'd require a 'report-type' replay.

Reply:
I added the `"!X 4X - no match" -> conflictsForNow(S0, F1, Nil),`. Making this and related cases merge is outside the scope of this PR

Reply:
I was not arguing to solve it in this pr.
We have a description of how the merging rules should behave but we are ignoring half of the data there. who will remember to pick up that work? if it's part of this compiling documentation - we have a spot that tracks it.

Reply:
Just to be clear:

> I don't think running the SM will be necessary. No match === no COMM event and this is all the info we need and have

It will. Notice we're talking about merging two deploys (and their events) that weren't executing together. E.g.:

I deploy `@0!(0)`
You deploy `for (_ <- @0) { @1(1) }`

None of the deploys will have a COMM (or a match) in their log, but we can't merge them, as they'd leave TS in undefined state (more work is needed).

Reply:
True